frcxx_nft/src/state.rs

```diff
@@ -171,6 +180,92 @@ impl NFTState {
         Ok(balance)
     }
 
+    /// Approves an operator to transfer a specific token
+    ///
+    /// Checks that the caller is the owner of the specified token
+    pub fn approve_for_token<BS: Blockstore>(
+        &mut self,
+        bs: &BS,
+        caller: ActorID,
+        operator: ActorID,
+        token_id: TokenID,
+    ) -> Result<()> {
+        let mut token_array = self.get_token_data_amt(bs)?;
+        let mut token_data = Self::owns_token(&token_array, caller, token_id)?;
+        token_data.operators.insert(operator);
+
+        token_array.set(token_id, token_data)?;
+        self.token_data = token_array.flush()?;
+
+        Ok(())
+    }
+
+    /// Revokes an operator to transfer a specific token
+    ///
+    /// Checks that the caller is the owner of the specified token
+    pub fn revoke_for_token<BS: Blockstore>(
+        &mut self,
+        bs: &BS,
+        token_id: TokenID,
+        caller: ActorID,
+        operator: ActorID,
+    ) -> Result<()> {
+        let mut token_array = self.get_token_data_amt(bs)?;
+        let mut token_data = Self::owns_token(&token_array, caller, token_id)?;
+        token_data.operators.remove(&operator);
+
+        token_array.set(token_id, token_data)?;
+        self.token_data = token_array.flush()?;
+
+        Ok(())
+    }
+
+    /// Approves an operator to transfer tokens on behalf of the owner
+    pub fn approve_for_all<BS: Blockstore>(
+        &mut self,
+        bs: &BS,
+        owner: ActorID,
+        operator: ActorID,
+    ) -> Result<()> {
+        let mut owner_map = self.get_owner_data_hamt(bs)?;
+        let new_owner_data = match owner_map.get(&actor_id_key(owner))? {
+            Some(data) => {
+                let mut operators = data.operators.clone();
+                operators.insert(operator);
+                OwnerData { operators, balance: data.balance }
+            }
+            None => OwnerData { balance: 0, operators: HashSet::new() },
+        };
+        owner_map.set(actor_id_key(owner), new_owner_data)?;
+        self.owner_data = owner_map.flush()?;
+
+        Ok(())
+    }
+
+    /// Revokes an operator's premission to transfer tokens on behalf of the owner
+    pub fn revoke_for_all<BS: Blockstore>(
```

Comment:
While `approve_for_all` fits nicely with the logic structure this name doesn't quite match the semantics I would expect given that you can call revoke_for_all for an operator but still have an approved operator transfer [if the token itself does not revoke that operator explicitly](https://github.com/helix-onchain/filecoin/pull/130/files#diff-d366e93293ca5febc8827cd9c7c8c52c4335d6b4b90e79e3c5f50ec8baa025d9R396).

Reply:
I've renamed it to `approve_for_owner` now but may change it again depending on discussion in the [FIP](https://github.com/filecoin-project/FIPs/discussions/492)

Ideally it'd just match the actor-interface method name